File: typescript-banco-digital/src/ContaPoupanca.ts
import { Conta } from "./Conta";

export class ContaPoupanca extends Conta {
    private _taxaJuros: number = 0.02; // 2%

    constructor(titular: string) {
        super(titular);
        console.log(`Conta Poupança de ${titular} criada com sucesso!`);
    }

    // Implementação obrigatória do método abstrato 'sacar'
    public sacar(valor: number): void {
        if (valor <= 0) {
            console.log("Valor de saque inválido.");
            return;
        }
        
        if (valor > this.getSaldo()) {
            console.log(`Saldo insuficiente. Saldo atual: R$${this.getSaldo().toFixed(2)}`);
        } else {
            this.setSaldo(this.getSaldo() - valor);
            console.log(`Saque de R$${valor.toFixed(2)} realizado. Saldo atual: R$${this.getSaldo().toFixed(2)}`);
        }
    }

    public calcularRendimento(): void {
        const rendimento = this.getSaldo() * this._taxaJuros;
        this.depositar(rendimento);
        console.log(`Rendimento de R$${rendimento.toFixed(2)} creditado.`);
    }
}

File: python-banco-digital/src/conta_poupanca.py
from conta import Conta

class ContaPoupanca(Conta):
    def __init__(self, titular: str, taxa_juros: float = 0.02):
        super().__init__(titular)
        self._taxa_juros = taxa_juros
        print(f"Conta Poupança de {titular} criada com sucesso!")

    # Implementação obrigatória do método abstrato
    def sacar(self, valor: float):
        if valor <= 0:
            print("Valor de saque inválido.")
            return
        
        if valor > self.saldo:
            print(f"Saldo insuficiente. Saldo atual: R${self.saldo:.2f}")
        else:
            self._saldo -= valor
            print(f"Saque de R${valor:.2f} realizado. Saldo atual: R${self._saldo:.2f}")

    def calcular_rendimento(self):
        rendimento = self.saldo * self._taxa_juros
        self.depositar(rendimento)
        print(f"Rendimento de R${rendimento:.2f} creditado.")